### PR TITLE
Efficient Sub-Pixel CNN for ADNI dataset

### PR DESCRIPTION
**Pull request by mikhail-matveev-V0**
<p>Added a PyTorch implementation of the the Efficient Sub-Pixel CNN, from the paper “Real-time single image and video super-resolution using an efficient sub-pixel convolutional neural network” [1], and applied it to the ADNI dataset [2].</p>

<p>Added four python files, a README.md which explains the model and its results, and an images folder which contains supporting images for README.md.</p>

<p>The added python files are as follows:</p>
<ul>
<li><b>modules.py</b> - contains the model definition</li>
<li><b>dataset.py</b> - contains code to load the training and testing dataset</li>
<li><b>train.py</b> - contains code to train the model defined in modules.py</li>
<li><b>predict.py</b> - contains example usage code of the trained model</li>
</ul>

The produced images, whilst not achieving perfect clarity, are much clearer and are a significant improvement over the downsampled images.

Author: Mikhail Matveev s4696671

<h4>Dependencies</h4>
<ul>
<li>Python 3.11</li>
<li>PyTorch 2.1.0</li>
<li>Matplotlib</li>
<li>ADNI dataset [2]</li>
</ul>

<h4>References</h4>
<p>[1] W. Shi, J. Caballero, F. Huszar, J. Totz, A. P. Aitken, R. Bishop, D. Rueckert, and Z. Wang, “Real-time
single image and video super-resolution using an efficient sub-pixel convolutional neural network,” in
<i>Proceedings of the IEEE Conference on Computer Vision and Pattern Recognition (CVPR)</i>, June 2016.</p>
<p>[2] The ADNI dataset for Alzheimer’s disease <a>https://adni.loni.usc.edu/</a></p>

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Overall this is a nice implementation. Training procedure seems ok with correct tracking of training loss each batch. Though no use of validation set. 

**[Recognition Problem]**
"modules.py":

- Correct implementation of PixelCNN, uses Tanh as-per the paper implementation. No changes to network architecturally.
- Uses a lambda function for pixel_shuffle (could have used the nn.module version rather than from functional but this shouldn't impact anything other than a bit of computation time).

"dataset.py":

- Nice usage of ImageLoader to deal with AD and NC folders.
- Adds some flipping for data augmentation.
- Rather large batch size (128).
- Uses a lambda function to call resize, could instead store the function to a variable and call that way, though again this shouldn't impact anything other than a bit of computation time.

"train.py":

- Doesn't split into training/validation.
- Only trains for 10 epochs, so won't print any metrics during training.
- It does track the batch-to-batch loss and plot afterwards.
- Uses OneCycleLR - seems to be correctly implemented.
- Does report MSE and PSNR on the test dataset after finished training (just stores the final model).
- Displays some sample images.

"predict.py":

- Simply loads a batch from the test dataset and uses them to generate a figure.
- Doesn't put the model in model.eval and still runs with grad. Shouldn't change results as no batch normalization or dropout is involved.
- Correctly loads model weights.
- Does not provide any metrics on the test dataset (but this is in "train.py").

"README.MD":

- Provides a brief overview of the network architecture.
- Model predictions are still a fair bit blurry (compared to other submissions, though some detail is added compared to low-rez).
- Reports dependencies.

**[Commit Log]**
Commit log looks ok.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage, but didnt use training validation splits -1
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
Please remove .gitignore file to allow merge. Doesnt affect grade, but only needed to complete merge.

**Comment by mikhail-matveev-V0** [comment]
> Please remove .gitignore file to allow merge. Doesnt affect grade, but only needed to complete merge.

Done!

**Comment by wangzhaomxy** [comment]
Feedback attempt and no feedback marks lost.
